src/js/libcs/OpDrawing.js

```diff
@@ -520,168 +534,131 @@ eval_helper.drawconic = function(conicMatrix, modifs) {
     var yRight = solveRealQuadratic( // x = csw
         c02, c11 * csw + c01, (c20 * csw + c10) * csw + c00);
     var xTop = solveRealQuadratic(c20, c10, c00); // y = 0
-    var xBottom = solveRealQuadratic(
+    var xBottom = solveRealQuadratic( // y = csh
         c20, c11 * csh + c10, (c02 * csh + c01) * csh + c00);
-    var tl = sign(0, 0);
-    var tr = sign(csw, 0);
-    var bl = sign(0, csh);
-    var br = sign(csw, csh);
-    if (!isFinite(tl * tr * bl * br))
-        return;
-    var boundary = [];
-
-    function doBoundary(sign1, sign2, sol, x, y, axis, sort, extent) {
-        var coord;
-        if (sign1 !== sign2) { // we need exactly one point of intersection
-            if (sol === null)
-                return false;
-            var center = extent * 0.5;
-            var dist0 = Math.abs(center - sol[0]);
-            var dist1 = Math.abs(center - sol[1]);
-            coord = sol[dist0 < dist1 ? 0 : 1];
-            boundary.push(mkp((1 - axis) * coord + x, axis * coord + y));
-        } else { // we need zero or two points of intersection
-            if (sol === null) // have zero intersections
-                return true;
-            coord = 0.5 * (sol[0] + sol[1]);
-            if (!(coord > 0 && coord < extent))
-                return true;
-            var signMid = sign((1 - axis) * coord + x, axis * coord + y);
-            if (signMid === sign1) // intersections outside segment
-                return true;
-            if (isNaN(signMid))
-                return true;
-            // Have two points of solution
-            sort = sort * (sol[1] - sol[0]) > 0 ? 0 : 1;
-            coord = sol[sort];
-            boundary.push(mkp((1 - axis) * coord + x, axis * coord + y));
-            coord = sol[1 - sort];
-            boundary.push(mkp((1 - axis) * coord + x, axis * coord + y));
+
+    // For ovals compute the roots of the x and y discriminant for
+    // the horizontal and vertical tangent points respectively
+    var flats = k00 <= 0 ? [null, null] : [
+        solveRealQuadratic(k00, -2 * k10, k20),
+        solveRealQuadratic(k00, -2 * k01, k02)
+    ];
+
+    var points = [];
+    // Inside state of first corner at top left
+    var bInside = (sign(0, 0) < 0) === (det < 0);
+
+    function doBoundary(sol, other, vert, sort, extent) {
+        // Edge cannot be inside if there are no solutions.
+        var t;
+        if (sol !== null) {
+            var coord, pt;
+            var extent0 = sort > 0 ? 0 : extent;
+            var extent1 = sort > 0 ? extent : 0;
+            // Only include each first corner that falls within the conic
+            if (bInside) {
+                t = "corner";
+                coord = extent0;
+                pt = vert ? mkp(other, coord, t) : mkp(coord, other, t);
+                points.push(pt);
+            }
+            coord = sort > 0 ? sol[0] : sol[1];
+            if (0 < coord && coord < extent || coord === extent1) {
```

Comment:
Relying on the inside or outside state of the first corner is definitely a bad idea. I'm currently rethinking this.